setup.py

```diff
@@ -54,6 +55,131 @@
+raw_extensions = [Extension("hyperspy.tests.misc.cython.test_cython_integration",
+                        ['hyperspy/tests/misc/cython/test_cython_integration.pyx']),
+                 ]
+
+cleanup_list = []
+for leftover in raw_extensions:
+    path, ext = os.path.splitext(leftover.sources[0])
+    if ext in ('.pyx', '.py'):
+        cleanup_list.append(os.path.join(setup_path, path + '.c*'))
+        if os.name == 'nt':
+            cleanup_list.append(os.path.join(setup_path, path + '.cpython-*.pyd'))
+        else:
+            cleanup_list.append(os.path.join(setup_path, path + '.cpython-*.so'))
+
+
+def count_c_extensions(extensions):
+    c_num = 0
+    for extension in extensions:
+        # if first source file with extension *.c or *.cpp exists
+        # it is cythonised or pure c/c++ extension:
+        sfile = extension.sources[0]
+        path, ext = os.path.splitext(sfile)
+        if os.path.exists(path + '.c') or os.path.exists(path + '.cpp'):
+            c_num += 1
+    return c_num
+
+
+def cythonize_extensions(extensions):
+    try:
+        from Cython.Build import cythonize
+        return cythonize(extensions)
+    except ImportError:
+        warnings.warn("""WARNING: cython required to generate fast c code is not found on this system.
+Only slow pure python alternative functions will be available.
+To use fast implementation of some functions writen in cython either:
+a) install cython and re-run the installation,
+b) try alternative source distribution containing cythonized C versions of fast code,
+c) use binary distribution (i.e. wheels, egg).""")
+        return []
+
+
+def no_cythonize(extensions):
+    for extension in extensions:
+        sources = []
+        for sfile in extension.sources:
+            path, ext = os.path.splitext(sfile)
+            if ext in ('.pyx', '.py'):
+                if extension.language == 'c++':
+                    ext = '.cpp'
+                else:
+                    ext = '.c'
+                sfile = path + ext
+            sources.append(sfile)
+        extension.sources[:] = sources
+    return extensions
+
+
+# to cythonize, or not to cythonize... :
+if len(raw_extensions) > count_c_extensions(raw_extensions):
+    extensions = cythonize_extensions(raw_extensions)
+else:
+    extensions = no_cythonize(raw_extensions)
+
+
+# HOOKS ######
+
+def find_post_checkout_cleanup_line():
+    """find the line index in the git post-checkout hooks"""
+    with open('.git/hooks/post-checkout', 'r') as pchook:
+        hook_lines = pchook.readlines()
+        for i in range(1, len(hook_lines), 1):
+            if re.search(r'^rm hyperspy/*', hook_lines[i]) is not None and \
+              (re.search(r'python.*?build_ext --inplace', hook_lines[i+1]) is not None):
+                return i
+
+# generate some git hook to clean up and re-build_ext --inplace
+# after changing branches:
+if os.path.exists('.git') and (not os.path.exists('.hook_ignore')):
```

Comment:
@sem-geologist This line above is what I meant - you should adopt the same 
`os.path.join(setup_path, '.git')` instead of just `'.git'` when checking if it exists.